File: backend/services/vector_store.py
"""
Pinecone vector store service.
Handles upsert and similarity search operations.
"""

import os
from pinecone import Pinecone, ServerlessSpec


EMBEDDING_DIM = 3072  # gemini-embedding-001 dimension


def _get_index():
    """Get or create the Pinecone index."""
    pc = Pinecone(api_key=os.getenv("PINECONE_API_KEY"))
    index_name = os.getenv("PINECONE_INDEX_NAME", "resume-screener")

    existing = {idx.name: idx for idx in pc.list_indexes()}

    if index_name in existing:
        # If dimension doesn't match, delete and recreate
        if existing[index_name].dimension != EMBEDDING_DIM:
            print(f"[Pinecone] Dimension mismatch — recreating index with dim={EMBEDDING_DIM}")
            pc.delete_index(index_name)
            import time
            time.sleep(5)  # wait for deletion to propagate
            existing = {}

    if index_name not in existing:
        pc.create_index(
            name=index_name,
            dimension=EMBEDDING_DIM,
            metric="cosine",
            spec=ServerlessSpec(cloud="aws", region="us-east-1"),
        )
        import time
        time.sleep(10)  # wait for index to be ready

    return pc.Index(index_name)


def add_documents(
    namespace: str,
    chunks: list[str],
    embeddings: list[list[float]],
    source_type: str,
) -> None:
    """
    Store document chunks and their embeddings in Pinecone.

    Args:
        namespace: Session-specific namespace for isolation.
        chunks: List of text chunks.
        embeddings: Corresponding embedding vectors.
        source_type: 'resume' or 'jd'.
    """
    index = _get_index()

    vectors = []
    for i, (chunk, embedding) in enumerate(zip(chunks, embeddings)):
        vectors.append({
            "id": f"{source_type}_{i}",
            "values": embedding,
            "metadata": {
                "text": chunk,
                "source": source_type,
                "chunk_index": i,
            },
        })

    # upsert in batches of 50
    batch_size = 50
    for i in range(0, len(vectors), batch_size):
        batch = vectors[i : i + batch_size]
        index.upsert(vectors=batch, namespace=namespace)

    print(f"[Pinecone] Stored {len(vectors)} vectors in namespace '{namespace}'")


def search_similar(
    namespace: str,
    query_embedding: list[float],
    top_k: int = 5,
) -> list[dict]:
    """
    Search for the most similar chunks in Pinecone.

    Args:
        namespace: Session namespace to search in.
        query_embedding: The query vector.
        top_k: Number of results to return.

    Returns:
        List of matches with text and score.
    """
    index = _get_index()

    results = index.query(
        vector=query_embedding,
        top_k=top_k,
        namespace=namespace,
        include_metadata=True,
    )

    matches = []
    for match in results.matches:
        matches.append({
            "text": match.metadata.get("text", ""),
            "score": match.score,
            "source": match.metadata.get("source", ""),
        })

    return matches

File: backend/src/services/vectorStore.ts
import { Pinecone, Index } from '@pinecone-database/pinecone';
import { ScoredResult } from '../types';
import dotenv from 'dotenv';
dotenv.config();

const pc = new Pinecone({ apiKey: process.env.PINECONE_API_KEY || '' });
const INDEX_NAME = process.env.PINECONE_INDEX_NAME || 'resume-screener';

let indexInstance: Index | null = null;

async function getIndex(): Promise<Index> {
  if (indexInstance) return indexInstance;

  const existingIndexes = await pc.listIndexes();
  const exists = existingIndexes.indexes?.some(idx => idx.name === INDEX_NAME);

  if (!exists) {
    await pc.createIndex({
      name: INDEX_NAME,
      dimension: 768, // Gemini text-embedding-004 output dimension
      metric: 'cosine',
      spec: { serverless: { cloud: 'aws', region: 'us-east-1' } },
    });
    // wait for index to be ready
    console.log('[Pinecone] Creating index, waiting for ready state...');
    await new Promise(resolve => setTimeout(resolve, 15000));
  }

  indexInstance = pc.index(INDEX_NAME);
  return indexInstance;
}

export async function addDocuments(
  sessionId: string,
  chunks: string[],
  embeddings: number[][],
  source: 'resume' | 'jd'
): Promise<void> {
  const index = await getIndex();
  const ns = index.namespace(sessionId);

  const vectors = chunks.map((text, i) => ({
    id: `${sessionId}_${source}_${i}`,
    values: embeddings[i],
    metadata: { text, source, chunkIndex: i },
  }));

  // upsert in batches of 100
  for (let i = 0; i < vectors.length; i += 100) {
    await ns.upsert(vectors.slice(i, i + 100));
  }
}

export async function searchSimilar(
  sessionId: string,
  queryEmbedding: number[],
  topK: number = 5
): Promise<ScoredResult[]> {
  const index = await getIndex();
  const ns = index.namespace(sessionId);

  const results = await ns.query({
    vector: queryEmbedding,
    topK,
    includeMetadata: true,
  });

  return (results.matches || []).map(match => ({
    document: {
      id: match.id,
      text: (match.metadata?.text as string) || '',
      embedding: [],
      metadata: {
        source: (match.metadata?.source as 'resume' | 'jd') || 'resume',
        chunkIndex: (match.metadata?.chunkIndex as number) || 0,
      },
    },
    score: match.score || 0,
  }));
}

export async function deleteSession(sessionId: string): Promise<void> {
  try {
    const index = await getIndex();
    await index.namespace(sessionId).deleteAll();
  } catch (_) {}
}
